components/scream/src/share/tests/tridiag_tests.cpp

```diff
@@ -0,0 +1,529 @@
+          continue;
+
+        // Skip combinations generated at this and higher levels that Solve::run
+        // doesn't support to reduce redundancies.
+        if ((nrhs  == 1 && data_pack_size > 1) ||
+            (nprob == 1 && A_pack_size    > 1))
+          continue;
+        if ((tc.solver == Solver::thomas_team_scalar ||
+             tc.solver == Solver::thomas_scalar ||
+             tc.solver == Solver::cr_scalar) &&
+            data_pack_size > 1)
+          continue;
+        if (static_cast<int>(APack::n) != static_cast<int>(DataPack::n) && nprob > 1)
+          continue;
+
+        const int prob_npack = npack<APack>(nprob);
+        const int rhs_npack = npack<DataPack>(nrhs);
+
+        TridiagArray<APack>
+          A("A", 3, nrow, prob_npack),
+          Acopy("A", A.extent(0), A.extent(1), A.extent(2));
+        DataArray<DataPack>
+          B("B", nrow, rhs_npack), X("X", B.extent(0), B.extent(1)),
+          Y("Y", X.extent(0), X.extent(1));
+        const auto Am = create_mirror_view(A);
+        const auto Bm = create_mirror_view(B);
+        {
+          const auto As = scalarize(Am);
+          const auto dl = subview(As, 0, ALL(), ALL());
+          const auto d  = subview(As, 1, ALL(), ALL());
+          const auto du = subview(As, 2, ALL(), ALL());
+          fill_tridiag_matrix(dl, d, du, nprob, nrhs /* seed */);
+        }
+        fill_data_matrix(scalarize(Bm), nrhs);
+        deep_copy(A, Am);
+        deep_copy(B, Bm);
+        deep_copy(Acopy, A);
+        deep_copy(X, B);
+
+        Solve<A_pack_size == data_pack_size, APack, DataPack>
+          ::run(tc, A, X, nprob, nrhs);
+
+        Real re; {
+          const auto Acopym = create_mirror_view(Acopy);
+          const auto As = scalarize(Acopym);
+          const auto dl = subview(As, 0, ALL(), ALL());
+          const auto d  = subview(As, 1, ALL(), ALL());
+          const auto du = subview(As, 2, ALL(), ALL());
+          const auto Xm = create_mirror_view(X);
+          const auto Ym = create_mirror_view(Y);
+          deep_copy(Acopym, Acopy);
+          deep_copy(Xm, X);
+          matvec(dl, d, du, scalarize(Xm), scalarize(Ym), nprob, nrhs);
+          re = reldif(scalarize(Bm), scalarize(Ym), nrhs);
+        }
+        const bool pass = re <= 50*std::numeric_limits<Real>::epsilon();
+        std::stringstream ss;
+        ss << Solver::convert(tc.solver) << " " << tc.n_kokkos_thread
+           << " " << tc.n_kokkos_vec << " | " << nrow << " " << nrhs << " "
+           << A_many << " | log10 reldif " << std::log10(re);
+        if ( ! pass) std::cout << "FAIL: " << ss.str() << "\n";
+        REQUIRE(pass);
+        //std::cout << "PASS: " << ss.str() << "\n";
+      }
+    }
+  }
+}
+
+template <int A_pack_size, int data_pack_size>
+void run_test () {
+  for (const auto solver : Solver::all) {
+    TestConfig tc;
+    tc.solver = solver;
+    if (scream::util::OnGpu<Kokkos::DefaultExecutionSpace>::value) {
+      tc.n_kokkos_vec = 1;
```

Comment:
Are you running 1 thread per warp here? Is it just to test robustness of the solver or is it something we may use on tridiag solvers? Why not testing say half a warp then?

Reply:
The solvers don't use Kokkos thread/vector loops; they get a h/w thread index. This is a check that indeed the solvers don't care about thread/vector config but just the product.